lisa/_assets/kmodules/sched_tp/perf_counters.c

```diff
@@ -0,0 +1,694 @@
+		hlist_del(&entry->group_link);
+		perfctr_event_release_entry(entry);
+	}
+	kfree(group);
+	return -ENOMEM;
+
+}
+
+/* Lookup match type */
+enum perfctr_match_type {
+	PERFCTR_MATCH_NAME,
+	PERFCTR_MATCH_STATUS
+};
+
+struct perfctr_match {
+	void *data;
+	enum perfctr_match_type type;
+};
+
+static int perfctr_event_activate(struct perfctr_core *perf_data,
+				  const struct perfctr_match *match)
+{
+	int result = -EINVAL, i;
+
+	struct perf_event_attr attr = {
+		.size		= sizeof(struct perf_event_attr),
+		.pinned		= 1,
+		.disabled	= 1,
+	};
+
+	for (i = 0; i < ARRAY_SIZE(perfctr_generic_lt); ++i) {
+		switch (match->type) {
+		case PERFCTR_MATCH_NAME:
+			if (strcmp(match->data, perfctr_generic_lt[i].name))
+				continue;
+			break;
+		case PERFCTR_MATCH_STATUS:
+			{
+				unsigned int status;
+
+				status = *(unsigned int *)match->data;
+				if (status != perfctr_generic_lt[i].default_on)
+					continue;
+				else
+					break;
+			}
+		default:
+			unreachable();
+		}
+		attr.config = perfctr_generic_lt[i].id;
+		attr.type   = perfctr_generic_lt[i].type;
+
+		mutex_lock(&perf_data->lock);
+		result = perfctr_event_activate_single(perf_data, &attr);
+		mutex_unlock(&perf_data->lock);
+		if (!result || match->type == PERFCTR_MATCH_NAME)
+			break;
+	}
+	return result;
+}
+
+static void perfctr_event_release_entry(struct perfctr_event_entry *entry)
+{
+	perf_event_disable(entry->event);
+	perf_event_release_kernel(entry->event);
+	kfree(entry);
+}
+
+static void perfctr_events_release_group(struct perfctr_core *perf_data,
+					  struct perfctr_event_group *group)
+{
+	struct perfctr_event_entry *entry;
+	struct hlist_node *next;
+
+	WARN_ON(!mutex_is_locked(&perf_data->lock));
+	hlist_for_each_entry_safe(entry, next, &group->entries, group_link) {
+		hlist_del_rcu(&entry->node);
+		synchronize_srcu(&perf_data->srcu);
+		hlist_del(&entry->group_link);
+		perfctr_event_release_entry(entry);
+	}
+	list_del(&group->node);
+	kfree(group);
+	--perf_data->nr_events;
+}
+
+static void perfctr_events_release(struct perfctr_core *perf_data)
+{
+	struct perfctr_event_group *group, *next;
+
+	mutex_lock(&perf_data->lock);
+	list_for_each_entry_safe(group, next, &perf_data->events, node) {
+		perfctr_events_release_group(perf_data, group);
+	}
+	mutex_unlock(&perf_data->lock);
+}
+
+#ifdef CONFIG_SYSFS
+static ssize_t pmu_raw_counters_show(struct kobject *kobj,
+				     struct kobj_attribute *attr, char *buf)
+{
+	struct perfctr_core *perf_data;
+	struct perfctr_event_group *group;
+	unsigned int count = 0;
+	ssize_t length = 0;
+
+	perf_data = container_of(kobj, struct perfctr_core, kobj);
+
+	mutex_lock(&perf_data->lock);
+	list_for_each_entry(group, &perf_data->events, node) {
+		length += sprintf(buf + length, "%d ", group->raw_id);
+		++count;
+	}
+	mutex_unlock(&perf_data->lock);
+
+	for(; count < perf_data->max_nr_events; ++count)
+		length += sprintf(buf + length, "%d ", HW_PERF_UNSUPPORTED);
+	length += sprintf(buf + length, "\n");
```

Comment:
I'm not sure about what is being printed here but it seems that we never do any bound checking on `buf` to ensure it's big enough